base-body/src/main/java/io/techery/janet/body/ActionBody.java

```diff
@@ -24,73 +17,33 @@ public ActionBody(String mimeType) {
         this.mimeType = mimeType;
     }
 
-    public abstract byte[] getContent() throws IOException;
-
-    private byte[] bytes() {
-        if (bytes == null) {
-            try {
-                bytes = getContent();
-                if (bytes == null) {
-                    throw new NullPointerException("bytes");
-                }
-            } catch (IOException e) {
-                e.printStackTrace();
-            }
-            if (bytes == null) bytes = new byte[0];
-        }
-        return bytes;
-    }
-
-    public String fileName() {
-        return null;
-    }
-
-    public String mimeType() {
+    public final String mimeType() {
         return mimeType;
     }
 
-    public InputStream in() throws IOException {
-        return new ByteArrayInputStream(bytes());
-    }
+    public abstract long length();
 
-    public long length() {
-        if (length > 0) {
-            return length;
-        }
-        length = bytes().length;
-        return length;
-    }
+    public abstract InputStream getContent() throws IOException;
 
-    public void writeTo(OutputStream out) throws IOException {
-        out.write(bytes());
-    }
+    public abstract void writeContentTo(OutputStream os) throws IOException;
 
     @Override public boolean equals(Object o) {
```

Comment:
incorrect the method `equals`  with `mimeType` field comparing. I recommend you delete this method to exclude this collision.